File: python/src/leetcode/reverse_integer.py
"""
Given a 32-bit signed integer, reverse digits of an integer.

Example 1:

Input: 123
Output:  321

Example 2:

Input: -123
Output: -321

Example 3:

Input: 120
Output: 21

Note:
Assume we are dealing with an environment which could only hold integers within
the 32-bit signed integer range. For the purpose of this problem, assume that
your function returns 0 when the reversed integer overflows.

https://leetcode.com/problems/reverse-integer/
"""

MAX_INT = 2 ** 31 - 1
MIN_INT = -2 ** 31


# noinspection PyPep8Naming,PyMethodMayBeStatic
class Solution:
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """
        if x > MAX_INT or x < MIN_INT:
            return 0

        s, sign = (str(x), 1) if x > 0 else (str(-x), -1)
        rev = sign * int(s[::-1])
        return 0 if rev > MAX_INT or rev < MIN_INT else rev

File: java/src/main/java/name/qidong/leetcode/ReverseInteger.java
package name.qidong.leetcode;

/**
 * Given a 32-bit signed integer, reverse digits of an integer.
 * <p>
 * <strong>Example 1:</strong>
 * <pre><code>
 * Input: 123
 * Output:  321
 * </code></pre>
 * <strong>Example 2:</strong>
 * <p>
 * <pre><code>
 * Input: -123
 * Output: -321
 * </code></pre>
 * <p>
 * <strong>Example 3:</strong>
 * <p>
 * <pre><code>
 * Input: 120
 * Output: 21
 * </code></pre>
 * <p>
 * <strong>Note:</strong> Assume we are dealing with an environment which could only hold integers
 * within the 32-bit signed integer range. For the purpose of this problem, assume that your
 * function returns 0 when the reversed integer overflows.
 *
 * @see <a href="https://leetcode.com/problems/reverse-integer/">Reverse Integer - LeetCode</a>
 */
public class ReverseInteger {
    public int reverse(int x) {
        if (x == Integer.MIN_VALUE) {
            return 0;
        }

        int sign = x < 0 ? -1 : 1;
        x *= sign;
        int unsigned = 0;
        while (x != 0) {
            int bits = calc_bits(unsigned);
            if (bits + 3 >= 32) {
                return 0;
            }

            unsigned *= 10;

            if (unsigned < 0) {
                return 0;
            }

            unsigned += x % 10;
            x /= 10;
        }
        return unsigned * sign;
    }

    private int calc_bits(int num) {
        int bits = 0;
        while (num != 0) {
            ++bits;
            num >>= 1;
        }
        return bits;
    }
}
